ca/keyreadwriter.go

```diff
@@ -0,0 +1,257 @@
+package ca
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/docker/swarmkit/ioutils"
+)
+
+const (
+	// keyPerms are the permissions used to write the TLS keys
+	keyPerms = 0600
+	// certPerms are the permissions used to write TLS certificates
+	certPerms = 0644
+)
+
+// KeyHeaderUpdater is a function which takes a set of PEM heaaders, the old
+// kek when writing, and the new kek when writing, and updates the PEM header
+// given both
+type KeyHeaderUpdater func(map[string]string, []byte, []byte) error
+
+// KeyReader reads a TLS cert and key from disk
+type KeyReader interface {
+	Read() ([]byte, []byte, error)
+	ReadHeaders() (map[string]string, []byte, error)
+	Target() string
+}
+
+// KeyWriter writes a TLS key and cert to disk
+type KeyWriter interface {
+	Write([]byte, []byte, *KEKUpdate) error
+	UpdateHeaders(cb func(map[string]string, []byte) error) error
+	RotateKEK(newKEK []byte) error
+	Target() string
+}
+
+// KEKUpdate provides an optional update to the kek when writing.  The structure
+// is needed so that we can tell the difference between "do not encrypt anymore"
+// and there is "no update".
```

Comment:
Yes, the KEK can potentially be nil.  If `KEKUpdate` is nil, there is no update.  If `KEKUpdate` is not nil, but the KEK is nil, we are rotating to having no KEK.